mitxgraders/stringgrader.py

```diff
@@ -43,34 +128,75 @@ def check_response(self, answer, student_input, **kwargs):
             answer (str): The answer to compare to
             student_input (str): The student's input passed by edX
         """
-        expect = answer['expect']
-        student = student_input
-
-        # What are we checking?
-        if self.config['accept_any']:
-            correct = True
-        elif self.config['accept_nonempty']:
-            correct = student.strip() != ""
+        expect = self.clean_input(answer['expect'])
+        student = self.clean_input(student_input)
+        incorrect_response = {'ok': False, 'grade_decimal': 0, 'msg': ''}
+
+        # Figure out if we are accepting any input,
+        # and construct the appropriate minimum length
+        accept_any = self.config['accept_any']
+        min_length = self.config['min_length']
+        if self.config['accept_nonempty']:
+            accept_any = True
+            if min_length == 0:
+                min_length = 1
+
+        # Apply the validation pattern
+        if self.config['validation_pattern'] is not None:
+            if not accept_any:
+                # Make sure that expect matches the pattern
+                match = re.match(self.config['validation_pattern'], expect)
+                if match is None:
+                    # expect doesn't match the required pattern,
+                    # so a student can never get this right.
+                    # Raise an error!
+                    msg = ("The provided answer '{}' does not match the validation pattern "
+                           "'{}'").format(answer['expect'], self.config['validation_pattern'])
+                    raise ConfigError(msg)
+
+            # Check to see if the student input matches the validation pattern
+            match = re.match(self.config['validation_pattern'], student)
+            if match is None:
+                # Either raise an error or grade as incorrect
+                if self.config['validation_err']:
+                    raise InvalidInput(self.config['invalid_msg'])
+                else:
+                    if self.config['debug']:
+                        incorrect_response['msg'] = 'Validation failed'
+                    return incorrect_response
+
+        # Perform the comparison
+        if accept_any:
+            # Just check the minimums. No comparison to expect.
+            # Check for the minimum length
+            chars = len(student)
+            if chars < min_length:
+                if self.config['explain_minimums']:
+                    msg = 'Your response is too short ({chars}/{min_length} characters)'
+                    incorrect_response['msg'] = msg.format(chars=chars,
+                                                           min_length=min_length)
+                return incorrect_response
```

Comment:
Question: Would it make more sense to raise an error here, so as not to use up student's attempt? Or make it configurable: `explain_minimums=Any('msg', 'error', None)`?

I'm not too familiar with the use case for `accept_any` so I'm not sure what the preference would be.

Reply:
The use case is typically "Write a sentence/paragraph on this topic", typically with unlimited submissions (because if you're accepting anything, does it really matter that they took a few submissions?). However, I do like your suggestion of raising an error rather than grading incorrect, or at least, having the option to do so. I'll implement it.

Reply:
The only downside I can think of is that errors don’t get logged for analytics (right?)

Reply:
So implemented.

Reply:
I'm not sure the error text gets logged; I think the submission does though (and you can tell that it wasn't an attempt-counting submission as attempts doesn't increment).